sycl/include/sycl/ext/intel/experimental/kernel_properties.hpp

```diff
@@ -10,72 +10,58 @@
+namespace experimental {
 
-// Implementation note: <property_class>::value fields should match property IDs
-// specified in llvm/lib/SYCLLowerIR/LowerKernelProps.cpp
-
-namespace detail {
-// Proxy to access private property classes' fields from the API code.
-template <class T> struct proxy {
-  static constexpr int value = T::value;
+enum class RegisterAllocMode : uint32_t {
+  AUTO = 0,
+  LARGE = 2,
 };
-} // namespace detail
 
-/// A boolean property which requests the compiler to use large register
-/// allocation mode at the expense of reducing the amount of available hardware
-/// threads.
-struct use_large_grf_tag {
-  template <class T> friend struct detail::proxy;
-
-private:
-  // Property identifier
-  static constexpr int value = 0;
+struct register_alloc_mode_key {
+  template <RegisterAllocMode Mode>
+  using value_t = oneapi::experimental::property_value<
+      register_alloc_mode_key, std::integral_constant<RegisterAllocMode, Mode>>;
 };
 
-__SYCL_DEPRECATED("use_double_grf is deprecated, use use_large_grf instead")
-inline constexpr use_large_grf_tag use_double_grf = {};
-inline constexpr use_large_grf_tag use_large_grf = {};
+template <RegisterAllocMode Mode>
+inline constexpr register_alloc_mode_key::value_t<Mode> register_alloc_mode;
+} // namespace experimental
+} // namespace intel
+
+namespace oneapi {
+namespace experimental {
 
-} // namespace kernel_properties
+template <>
+struct is_property_key<intel::experimental::register_alloc_mode_key>
+    : std::true_type {};
 
-namespace __MP11_NS = sycl::detail::boost::mp11;
+namespace detail {
+template <>
+struct PropertyToKind<intel::experimental::register_alloc_mode_key> {
+  static constexpr PropKind Kind = PropKind::RegisterAllocMode;
+};
 
-// TODO this should be replaced with the generic SYCL compile-time properites
-// mechanism once implementaion is available.
-// https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_properties.asciidoc
+template <>
+struct IsCompileTimeProperty<intel::experimental::register_alloc_mode_key>
+    : std::true_type {};
 
-template <class... KernelProps>
-void set_kernel_properties(KernelProps... props) {
-  // TODO check for duplicates
-  using Props = __MP11_NS::mp_list<KernelProps...>;
-  __MP11_NS::mp_for_each<Props>([&](auto Prop) {
-    using PropT = decltype(Prop);
-    constexpr bool IsLargeGRF =
-        std::is_same_v<PropT, kernel_properties::use_large_grf_tag>;
-    if constexpr (IsLargeGRF) {
-      __sycl_set_kernel_properties(
-          kernel_properties::detail::proxy<
-              kernel_properties::use_large_grf_tag>::value);
-    } else {
-      static_assert(IsLargeGRF &&
-                    "set_kernel_properties: invalid kernel property");
-    }
-  });
-}
+template <intel::experimental::RegisterAllocMode Mode>
+struct PropertyMetaInfo<
+    intel::experimental::register_alloc_mode_key::value_t<Mode>> {
+  static constexpr const char *name = "RegisterAllocMode";
```

Comment:
Other properties try to follow the `sycl-prop` where `prop` is the name of the property with `_` replaced with `-`. In this case it would be `sycl-register-alloc-mode`. Would it be alright to stay in line with that?

Reply:
yes we can do that, done in latest commit